Python/bltinmodule.c

```diff
@@ -2489,13 +2489,72 @@ This function is intended specifically for use with numeric values and may
 reject non-numeric types.
 [clinic start generated code]*/
 
+static PyObject* range_sum_fastpath(PyObject* module, PyObject *range)
+{
+    PyObject* length = builtin_len(module, range);
+
+    if (PyObject_RichCompareBool(length, PyLong_FromLong(0), Py_EQ)) {
+        Py_DecRef(length);
+        return PyLong_FromLong(0);
+    }
+
+    // compute start * length + step * length * (length - 1) // 2
+    //                                          [    a     ]
+    //                                 [         b         ]
+    //                                 [             c           ]
+    //                         [                 d               ]
+    //        [       e      ]
+    //        [                      result                      ]
+
+    PyObject* start = PyObject_GetAttrString(range, "start");
+    PyObject* step = PyObject_GetAttrString(range, "step");
+
+    PyObject* one = PyLong_FromLong(1);
+    PyObject* a = PyNumber_Subtract(length, one);
+
+    PyObject* b = PyNumber_Multiply(a, length);
+
+    PyObject* two = PyLong_FromLong(2);
+    PyObject* c = PyNumber_FloorDivide(b, two);
+
+    PyObject* d = PyNumber_Multiply(step, c);
+
+    PyObject* e = PyNumber_Multiply(length, start);
+
+    PyObject* result = PyNumber_Add(d, e);
```

Comment:
All of these calls can return `NULL`. You should prevent these situations.

Reply:
To be clear, do you mean adding something like:

```c
if (one == NULL) {
    Py_DECREF(one);
    return NULL;
}
```

after every one of these? I would have to do it after 

```c
        PyObject* rangesum =  range_sum_fastpath(module, iterable);
        result = PyNumber_Add(result, rangesum);
        Py_DECREF(rangesum);
        return result;
```

as well then, I think.